feat(gatsby-theme-bodiless): Bodiless Image Refresh

beliayeu opened the pull request:
<!--
  Before issuing a pull request:

  - Please read our contribution guidelines (https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/packages/bodiless-documentation/doc/Development/Contributing.md), especially the section on Pull Requests.
  - Please first create an issue (https://github.com/johnsonandjohnson/Bodiless-JS/issues/new). All pull requests must be linked to an issue.
-->

<!--
  IMPORTANT
  
  The pull request title will become the commit message title at merge, and
  should adhere to Angular Commit Message Conventions. Please see (https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/packages/bodiless-documentation/doc/Development/Contributing.md) for details and examples.
-->

## Changes
* added a mechanism to enable gatsby image for images with default content
  * implemented 2 options how site builders can configure default content sources. The first options is to specify sources in site gatsby-config.js. The second option is to enable default content auto-discovery via environment variable.
* configured gatsby image for default content touts and images on test-site
* migrated data for images that have "Alt Text" value in their json, replaced "Alt Text" with empty string
* added `asDesignableGatsbyImage` that allows customizing Image and GatsbyImage components leveraging design API
* added `withDefaultImageContent` to @bodiless/components
* added `withImageContentLibrary` to @bodiless/components
* extended `asToken` so that it allows accepting undefined as an argument

## Notes
* activated gatsby image for touts on test-site, but disabled it for horizontal tout rendered on /touts page so that do not break existing cypress test for touts. gatsby image can be enabled for horizontal tout back once the cypress test is refactore

## Test Instructions
<!-- Detailed description of how this feature was (and should be) tested
  - Setup instructions (if any)?
  - List of test cases, with steps and expected results?
  - List
-->

## Related Issues

Closes #852
Related to #798

<!--
  Link to the issue that is fixed or resolved by this PR (if there is one)
  e.g. Fixes #1234, Closes #4567

  Link to an issue that is partially addressed by this PR (if there are any)
-->

<!-- IF THIS PR INTRODUCES A BREAKING CHANGE

BREAKING CHANGE: Describe the nature of the breaking change here.

More Details about the breaking change.
-->

wodenx commented on packages/bodiless-layouts/src/ContentLibrary/withContentLibrary.tsx:
```diff
@@ -40,7 +40,7 @@ const copyNode = (source: ContentNode<any>, dest: ContentNode<any>, copyChildren
 
 const withContentLibrary = (options: ContentLibraryOptions) => (
-  Component: ComponentType,
+  Component: ComponentType | string,
```
again let's type it as ComponentOrTag

beliayeu replied:
fixed in 6e2b015da42a9793abb10452f5eac71a95e6da7c

wodenx commented on packages/bodiless-core/src/Contentful/ContentfulNode.ts:
```diff
@@ -50,10 +50,12 @@ export default class ContentfulNode<D extends object> extends DefaultContentNode
     const contentKey = this.getContentKey();
-    return (this.content as any)[contentKey] || {};
+    const contentValue = (this.content as any)[contentKey];
+    if (typeof contentValue === 'function') return contentValue(this.proxy({}));
```
Why are we returning a proxy with no handlers?

beliayeu replied:
I believe we do not need proxy here
removed in 51f9085cb7e8edb4b82ceb24d9bd06c40a289935

beliayeu commented on examples/test-site/gatsby-config.js:
```diff
@@ -1,4 +1,11 @@
+const {
+  createDefaultContentPlugins,
+  getSampleDefaultContentConfig,
+} = require('@bodiless/gatsby-theme-bodiless/dist/DefaultContent');
```
require from '@gatsby-theme-bodiless' fails because of require of assets made by gatsby-theme-bodiless. particularly, bodiless-ui/lib/Spinner.css.
require from '@gatsby-theme-bodiless' works from site components as far as the execution is made in webpack context which knows how to load them
here, we are not in webpack context and asset require will throw an error

beliayeu commented on packages/gatsby-theme-bodiless/src/dist/DefaultContent/getSampleConfig.ts:
```diff
@@ -0,0 +1,27 @@
+import path from 'path';
+
+const getConfig = () => [
+  path.resolve(
```
I believe we can use require.resolve to find a path to @bodiless/gatsby-theme-bodiless
but I think we will still need doing path.resolve

```
const getConfig = () => [
  path.resolve(
    path.dirname(
      require.resolve('@bodiless/gatsby-theme-bodiless', { paths: [process.cwd()] }),
    ),
    'dist',
    'DefaultContent',
    'sample',
  ),
];
```

beliayeu commented on examples/test-site/gatsby-config.js:
```diff
@@ -36,6 +43,10 @@ const plugins = [
     resolve: 'gatsby-plugin-sitemap',
   },
+  ...createDefaultContentPlugins(
+    ...getSampleDefaultContentConfig(),
```
the problem here is that the options we pass in site's gatsby-config.js won't be available in a plugin gatsby-config.js, particularly, we can not pass configuration from site gatsby-config.js and make it available in @bodiless/gatsby-theme-bodiless gatsby-config

per discussion in skype, decided not to spend more time finding a way how to pass data from site to gatsby-theme-bodiless as far as mainly autodiscovery will be enabled on sites and this configuration won't be needed

beliayeu commented on packages/bodiless-components/src/Image/withDefaultImageContent.ts:
```diff
@@ -0,0 +1,40 @@
+  placeholder,
+  useOverrides,
+) => asToken(
+  asEditableImage.meta !== undefined ? asEditableImage.meta : identity,
```
makes sense, meta is added to hocs, while AsBodiless is not a hoc, it returns a hoc
updated in d21922b1935a3a48224f52806fd44be6bcaea55a
also, made asToken accept undefined and validated that asToken always attaches empty meta property

beliayeu commented on packages/bodiless-components/src/Image/withImageLibrary.ts:
```diff
@@ -0,0 +1,64 @@
+
+// Adds image library to an asEditableImage hoc.
+const withImageLibrary = (
+  asEditableImage: AsBodilessImage & { meta?: TokenMeta },
```
fixed in d21922b

beliayeu commented on packages/bodiless-core/src/Contentful/ContentfulNode.ts:
```diff
@@ -35,7 +35,7 @@ export default class ContentfulNode<D extends object> extends DefaultContentNode
 
   // @ts-ignore has no initializer and is not definitely assigned in the constructor
-  private content: D;
+  private content: object;
```
fixed in 6e2b015da42a9793abb10452f5eac71a95e6da7c

beliayeu commented on packages/bodiless-core/src/Contentful/withDefaultContent.tsx:
```diff
@@ -35,7 +35,7 @@ import { DefaultContentNode } from '../ContentNode';
  */
 const withDefaultContent = <P extends object, D extends object>(content: D|((props:P) => D)) => (
-  (Component: CT<P>) => {
+  (Component: CT<P> | string) => {
```
fixed in 6e2b015da42a9793abb10452f5eac71a95e6da7c

beliayeu commented on packages/bodiless-core/src/asBodilessComponent.tsx:
```diff
@@ -16,6 +16,7 @@ import React, { ComponentType as CT } from 'react';
 import {
   pick, omit, identity, flowRight,
 } from 'lodash';
+import type { HOC } from '@bodiless/fclasses/lib/Tokens';
```
agreed, the reason I did was that there was an existing HOC type imported from package root
tried to replace, made the new HOC type exported by fclasses, no ts errors found
the fix is made in 6e2b015da42a9793abb10452f5eac71a95e6da7c

beliayeu replied:
the new HOC type introduces build issue on an existing site - discuss if we want to accept this breaking change

beliayeu commented on packages/gatsby-theme-bodiless/gatsby-config.js:
```diff
@@ -101,6 +104,40 @@ const getbuildCSSPlugins = require('./build-css');
+    ]);
+    // eslint-disable-next-line no-loop-func
+    files.forEach(file => {
+      const fileContent = JSON.parse(fs.readFileSync(file));
```
made experiments with making the logic async
wip: https://github.com/beliayeu/Bodiless-JS/compare/feat/bodiless-image-refresh...beliayeu:feat/bodiless-image-refresh-async-discovery?expand=1

it looks there is a constraint that we can not use async/await in gatsby-config
on the option we can do is to prebuild the file containing configuration

related discussion in gatsby: https://github.com/gatsbyjs/gatsby/issues/11128 and https://github.com/gatsbyjs/gatsby/issues/19644

beliayeu commented on packages/gatsby-theme-bodiless/gatsby-config.js:
```diff
@@ -101,6 +104,40 @@ const getbuildCSSPlugins = require('./build-css');
+ * default content plugins
+ */
+const discoverDefaultContent = (depth = 1) => {
+  let dir = path.resolve(process.cwd());
```
for get it working in a monorepo, there is an env variable `BODILESS_DEFAULT_CONTENT_AUTO_DISCOVERY_DEPTH` that allows to configure amount of dirs to navigate up

beliayeu commented on packages/gatsby-theme-bodiless/src/dist/GatsbyImage/asGatsbyImage.tsx:
```diff
@@ -57,6 +57,9 @@ const asGatsbyImage$ = (Component: CT<any>) => {
 
+/**
+ * @deprecated in favor of asDesignableGatsbyImage
+ */
```
I believe we can merge it, tried the new version on one of our sites, did not get any issues
merged asDesignableGatsbyImage into asGatsbyImage in bba137b42ff6fc0b9f33bbeba80bda24c11a131f

wodenx commented on packages/bodiless-components/src/Image/Image.tsx:
```diff
@@ -211,7 +213,8 @@ const options: BodilessOptions<Props, Data> = {
 
-export type AsBodilessImage = AsBodiless<HTMLProps<HTMLImageElement>, Data>;
+export type AsBodilessImage = AsBodiless<ImageProps, Data>;
+export type AsImageToken = ReturnType<AsBodilessImage> & TokenProps<ImageProps>;
```
I think this should be named just "ImageToken"

beliayeu replied:
since Token is included to AsBodiless
removed ImageToken in b74b8b0877779b76fa35b2d0627d8a24aa507fb8

wodenx commented on packages/bodiless-components/src/Image/withDefaultImageContent.ts:
```diff
@@ -14,27 +14,29 @@
-  typeof nodeKey === 'string' ? withResetButton({ nodeKey }) : withResetButton(nodeKey),
-);
+) => {
+  const asImageHoc = asEditableImage(nodeKey, placeholder, useOverrides) as AsImageToken;
```
Why is this cast necessary?   AsEditableImage should return an ImageToken

beliayeu replied:
currently, AsEditableImage has a specific type of AsBodiless generic, so it returns HOC.
to my mind we have 2 options:
- either to include Token into AsBodiless ReturnType, or
- allow asEditableImage be HOC or Token. Then we will need a way how to determine whether a particular function is a hoc or token. As an option we can add a property to token.

wodenx commented on packages/bodiless-components/src/Image/withDefaultImageContent.ts:
```diff
@@ -14,27 +14,29 @@
+    typeof nodeKey === 'string' ? withDefaultContent({
+      [nodeKey]: nodeContent,
+    }) : identity,
+    typeof nodeKey === 'string' ? withResetButton({ nodeKey }) : withResetButton(nodeKey),
```
Seems like we should move this overload to withResetButton...

beliayeu replied:
fixed in b74b8b0877779b76fa35b2d0627d8a24aa507fb8

wodenx commented on packages/bodiless-layouts/src/ContentLibrary/withContentLibrary.tsx:
```diff
@@ -40,7 +41,7 @@ const copyNode = (source: ContentNode<any>, dest: ContentNode<any>, copyChildren
 
 const withContentLibrary = (options: ContentLibraryOptions) => (
-  Component: ComponentType | string,
+  Component: ComponentOrTag<any>,
```
Should we make this generic?
```
const withContentLibrary = (options: ContentLIbraryOptions) => (
  <P extends object>(Component: ComponentOrTag<P>) => ...

beliayeu replied:
fixed in b74b8b0877779b76fa35b2d0627d8a24aa507fb8

wodenx commented on packages/fclasses/src/Tokens.tsx:
```diff
@@ -70,7 +70,7 @@ type TokenFilterTest<P> = (hoc: Token<P>) => boolean;
  * objects in addition to tokens.
  */
-type TokenDef<P> = Token<P>|TokenMeta;
+type TokenDef<P> = Token<P>|TokenMeta|undefined;
```
I have this change in my pr, but i think it also requires [dealing with undefined tokendefs in asToken](https://github.com/johnsonandjohnson/Bodiless-JS/pull/906/files#diff-c413d4b2772969b0166ede9ccc9c509e543323df95e6d9c6241f96a2a5057a6aR190-R192).  Do you need this "undefined" in this PR? or can we wait for it in mine?

beliayeu replied:
yep, I need it, since there is an access to meta prop which can undefined in image hocs
for example, in withDefaultImageContent

TatsianaMiraniuk commented:
Hi @beliayeu ,

please see **issue_1**:

there are 2 warnings in a console in Test Site, on each page (related to Logo image) - Edit env only:
1) Warning: React does not recognize the `gatsbyImg` prop on a DOM element. ...
2) Warning: React does not recognize the `publicUrl` prop on a DOM element. ..

TatsianaMiraniuk commented:
@beliayeu ,

please see **issue_2:**

An uploaded image is in low quality, e.g. upload an image in Hero image on Home page or in a Tout on /touts/ page. 
Please see screenshot to compare with the current master behavior:
![Screen Shot 2021-03-02 at 5 40 38 PM](https://user-images.githubusercontent.com/3224040/109725348-ff0bef00-7b7e-11eb-97c2-5af332581044.png)

TatsianaMiraniuk commented:
@beliayeu ,

please see **issue_3:**
The images are not set for Contentful Touts that are chosen from Component Picker.

### Steps to reproduce:
1. open Edit env, /flow-container/ page
2. switch to Edit mode
3. click FC 
4. click 'Add'
5. select Contentful Tout in the filter
### Actual Result:
no image is set
![Screen Shot 2021-03-02 at 5 52 16 PM](https://user-images.githubusercontent.com/3224040/109726354-9b82c100-7b80-11eb-9c22-a32168f8758b.png)


6. select any Contentful Tout
### Actual Result:
no image is shown when the component is put on the page.

### Expected Results:
Contentful Tout has pre-set content (including an image)

beliayeu commented:
thanks, @TatsianaMiraniuk 

issue_1:
problem:
image data is extended during gatsby node creation phase with additional data required for default content case.
standard image hoc does not perform props stripping, so we ended up with invalid html attributes

fix:
provided a new hoc on site level that can used for rendering non-responsive image, this hoc performs props stripping
activated this new hoc for logo

fixed in 3272da60d2deba467aaada56f2fa67f3dcb6d4e2

issue_2: I think that is a trade-off between image quality and image size.
to discuss with team whether we want to adjust image quality on test-site
cc: @hvanyo 

issue3:
problem: default data is not available for this page
fix: added DefaultContentQuery graphql query so that default data is available

fixed in 3272da60d2deba467aaada56f2fa67f3dcb6d4e2

TatsianaMiraniuk commented:
@beliayeu ,

please see **issue_4:**:

there is design="[object Object]" for the placeholder images in Touts:
![Screen Shot 2021-03-03 at 3 00 16 PM](https://user-images.githubusercontent.com/3224040/109864790-494da880-7c31-11eb-8959-f0da4e0b88d8.png)

TatsianaMiraniuk commented:
@beliayeu ,

please see **issue_5**:
Image path from Content Library isn't pulled correctly in Src field of Image Upload form

### Steps to reproduce:
1. in Edit env, open /content-library/
2. switch to Edit mode
3. in Image with library, select any image from Library
4. open Image Upload form (click Image: Select)
### Actual Result:
Src value isn't showing the correct path
![Screen Shot 2021-03-03 at 3 26 36 PM](https://user-images.githubusercontent.com/3224040/109868090-3d63e580-7c35-11eb-9c78-afee5fec63b0.png)

4. click V

### Actual Result:
the image doesn't show (due to the incorrect path)
![Screen Shot 2021-03-03 at 3 30 55 PM](https://user-images.githubusercontent.com/3224040/109868315-78661900-7c35-11eb-9323-ff25e50613e6.png)

TatsianaMiraniuk commented:
@beliayeu ,

please see **issue_6:**
The issue is similar to **issue_5** but it causes the env to return 502 error

Image path isn't pulled correctly in Src field of Image Upload form

### Steps to reproduce:
1. in Edit env, open /images/
2. switch to Edit mode
3. in Contentful Image Components, open Image Upload Form
4. click V

### Actual Result:
the env returns 502 error

TatsianaMiraniuk commented:
@beliayeu ,

please see **issue_7**:
Images in 'Gatsby Images' section on /images/ page are not Gatsby images.
![Screen Shot 2021-03-11 at 5 12 38 PM](https://user-images.githubusercontent.com/3224040/110862297-51d75c00-828d-11eb-9e4a-dc6ca5996389.png)

TatsianaMiraniuk commented:
**issue_8:**
Gatsby images show incorrectly, no padding  in /images/ page for <1024px
![Screen Shot 2021-03-12 at 5 20 28 PM](https://user-images.githubusercontent.com/3224040/111160579-9b1afa80-8570-11eb-81a4-750ea024a83e.png)
![Screen Shot 2021-03-12 at 5 20 55 PM](https://user-images.githubusercontent.com/3224040/111160587-9e15eb00-8570-11eb-8f96-7701af26cb90.png)

TatsianaMiraniuk commented:
Hi @beliayeu ,

please see issue_9, specific for Starter Site:

**Steps to reproduce:**
1. open Home page in Edit env for 1024+ px
2. upload Hero image
3. upload images in Touts
4. refresh the page
5. resize the browser to less than 1024px

**Actual Results:**
Hero image, Touts get much smaller:
![Screen Shot 2021-03-16 at 1 51 11 PM](https://user-images.githubusercontent.com/3224040/111373584-73ac5680-8672-11eb-9a9c-88aebe6c00d4.png)

6. switch to Preview Mode

**Actual Results:**
Hero image is missing, Touts are small
![Screen Shot 2021-03-16 at 1 49 36 PM](https://user-images.githubusercontent.com/3224040/111373648-8d4d9e00-8672-11eb-8001-e514cadcad43.png)
Same in Static:
![Screen Shot 2021-03-16 at 4 05 21 PM](https://user-images.githubusercontent.com/3224040/111373717-a6564f00-8672-11eb-9be9-14735fa436cf.png)
